spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/TaskLifecycleListener.java

```diff
@@ -0,0 +1,162 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.task.listener;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Date;
+import java.util.UUID;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.beans.BeansException;
+import org.springframework.boot.context.event.ApplicationFailedEvent;
+import org.springframework.cloud.task.repository.TaskExecution;
+import org.springframework.cloud.task.repository.TaskRepository;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.event.ContextClosedEvent;
+import org.springframework.context.event.ContextRefreshedEvent;
+
+/**
+ * Monitors the lifecycle of a task.  This listener will record both the start and end of
+ * a task in the registered {@link TaskRepository}.
+ *
+ * The following events are used to identify the start and end of a task:
+ *
+ * <ul>
+ *     <li>{@link ContextRefreshedEvent} - Used to identify the start of a task.  A task
+ *     is expected to contain a single application context.</li>
+ *     <li>{@link ContextClosedEvent} - Used to identify the successful end of a task.</li>
+ *     <li>{@link ApplicationFailedEvent} - Used to identify the failure of a task.</li>
+ * </ul>
+ *
+ * <b>NOTE:</b> Multiple contexts (including parent/child relationships) will result in
+ * only the first context refresh that contains this instance being recorded.
+ *
+ * @author Michael Minella
+ */
+public class TaskLifecycleListener implements ApplicationListener<ApplicationEvent>,
+		ApplicationContextAware {
+
+	private final static Logger logger = LoggerFactory.getLogger(TaskLifecycleListener.class);
+
+	private final TaskRepository taskRepository;
+
+	private TaskExecution taskExecution;
+
+	private boolean started = false;
+
+	private ApplicationContext applicationContext;
+
+	/**
+	 * @param taskRepository The repository to record executions in.
+	 */
+	public TaskLifecycleListener(TaskRepository taskRepository) {
+		this.taskRepository = taskRepository;
+	}
+
+	/**
+	 * @param applicationEvent The application being listened for.
+	 */
+	@Override
+	public void onApplicationEvent(ApplicationEvent applicationEvent) {
+		if(applicationEvent instanceof ContextRefreshedEvent) {
+			doTaskStart();
+			started = true;
+		}
+		else if(applicationEvent instanceof ContextClosedEvent) {
+			doTaskEnd();
+		}
+		else if(applicationEvent instanceof ApplicationFailedEvent) {
+			doTaskFailed(((ApplicationFailedEvent) applicationEvent).getException());
+		}
+	}
+
+	private void doTaskFailed(Throwable exception) {
+
+		if(started) {
+			this.taskExecution.setEndTime(new Date());
+			//TODO: get exit code from new event thrown by Boot.
```

Comment:
Agreed.  Really depends on the implementation that boot will be adding on uncaught exceptions.  This comment is based on the current implementation (which is going away)